### PR TITLE
[security_monitoring_rule] Support Third Party rules

### PR DESCRIPTION
**Pull request by pietrodll**
Add support for Cloud SIEM [third party detection rules](https://docs.datadoghq.com/security/cloud_siem/log_detection_rules/?tab=thirdparty#third-party).

- Bump `datadog-api-client-go` to include https://github.com/DataDog/datadog-api-client-go/pull/2337
- Add `third_party_case`
- Add `third_party_rule_options`
- Add utility function `parseStringArray` to convert `[]interface{}` to `[]string`

### CODE REVIEW

**Review by daniel-trinh — APPROVED** [review]

File: docs/resources/security_monitoring_rule.md [hunk 1]
```diff
@@ -54,19 +54,20 @@ resource "datadog_security_monitoring_rule" "myrule" {
...
 
+- `case` (Block List, Max: 10) Cases for generating signals. (see [below for nested schema](#nestedblock--case))
 - `enabled` (Boolean) Whether the rule is enabled. Defaults to `true`.
 - `filter` (Block List) Additional queries to filter matched events before they are processed. (see [below for nested schema](#nestedblock--filter))
 - `has_extended_title` (Boolean) Whether the notifications include the triggering group-by values in their title. Defaults to `false`.
 - `options` (Block List, Max: 1) Options on rules. (see [below for nested schema](#nestedblock--options))
 - `query` (Block List) Queries for selecting logs which are part of the rule. (see [below for nested schema](#nestedblock--query))
 - `signal_query` (Block List) Queries for selecting logs which are part of the rule. (see [below for nested schema](#nestedblock--signal_query))
 - `tags` (Set of String) Tags for generated signals.
+- `third_party_case` (Block List, Max: 10) Cases for generating signals for third party rules. Only required and accepted for third party rules (see [below for nested schema](#nestedblock--third_party_case))
```
Comment (brett0000FF) [inline docs/resources/security_monitoring_rule.md:70]:
```suggestion
- `third_party_case` (Block List, Max: 10) Cases for generating signals for third-party rules. Only required and accepted for third-party rules (see [below for nested schema](#nestedblock--third_party_case))
```

File: docs/resources/security_monitoring_rule.md [hunk 2]
```diff
@@ -99,18 +100,16 @@ Required:
...
 - `impossible_travel_options` (Block List, Max: 1) Options for rules using the impossible travel detection method. (see [below for nested schema](#nestedblock--options--impossible_travel_options))
+- `keep_alive` (Number) Once a signal is generated, the signal will remain “open” if a case is matched at least once within this keep alive window (in seconds). Valid values are `0`, `60`, `300`, `600`, `900`, `1800`, `3600`, `7200`, `10800`, `21600`.
+- `max_signal_duration` (Number) A signal will “close” regardless of the query being matched once the time exceeds the maximum duration (in seconds). This time is calculated from the first seen timestamp. Valid values are `0`, `60`, `300`, `600`, `900`, `1800`, `3600`, `7200`, `10800`, `21600`, `43200`, `86400`.
 - `new_value_options` (Block List, Max: 1) New value rules specific options. (see [below for nested schema](#nestedblock--options--new_value_options))
+- `third_party_rule_options` (Block List, Max: 1) Options for rules using the third party detection method. (see [below for nested schema](#nestedblock--options--third_party_rule_options))
```
Comment (brett0000FF) [inline docs/resources/security_monitoring_rule.md:112]:
```suggestion
- `third_party_rule_options` (Block List, Max: 1) Options for rules using the third-party detection method. (see [below for nested schema](#nestedblock--options--third_party_rule_options))
```

File: docs/resources/security_monitoring_rule.md [hunk 3]
```diff
@@ -134,6 +133,32 @@ Optional:
 - `learning_threshold` (Number) A number of occurrences after which signals are generated for values that weren't learned. Valid values are `0`, `1`. Defaults to `0`.
 
 
+<a id="nestedblock--options--third_party_rule_options"></a>
+### Nested Schema for `options.third_party_rule_options`
+
+Required:
+
+- `default_status` (String) Severity of the default rule case, when none of the third party cases match. Valid values are `info`, `low`, `medium`, `high`, `critical`.
```
Comment (brett0000FF) [inline docs/resources/security_monitoring_rule.md:141]:
```suggestion
- `default_status` (String) Severity of the default rule case, when none of the third-party cases match. Valid values are `info`, `low`, `medium`, `high`, `critical`.
```

File: docs/resources/security_monitoring_rule.md [hunk 4]
```diff
@@ -134,6 +133,32 @@ Optional:
...
+- `root_query` (Block List, Min: 1, Max: 10) Queries to be combined with third party case queries. Each of them can have different group by fields, to aggregate differently based on the type of alert. (see [below for nested schema](#nestedblock--options--third_party_rule_options--root_query))
```
Comment (brett0000FF) [inline docs/resources/security_monitoring_rule.md:142]:
```suggestion
- `root_query` (Block List, Min: 1, Max: 10) Queries to be combined with third-party case queries. Each of them can have different group by fields, to aggregate differently based on the type of alert. (see [below for nested schema](#nestedblock--options--third_party_rule_options--root_query))
```

File: docs/resources/security_monitoring_rule.md [hunk 5]
```diff
@@ -134,6 +133,32 @@ Optional:
...
+
+Optional:
+
+- `default_notifications` (List of String) Notification targets for the default rule case, when none of the third party cases match.
```
Comment (brett0000FF) [inline docs/resources/security_monitoring_rule.md:146]:
```suggestion
- `default_notifications` (List of String) Notification targets for the default rule case, when none of the third-party cases match.
```

File: docs/resources/security_monitoring_rule.md [hunk 6]
```diff
@@ -177,6 +202,20 @@ Optional:
...
+
+Required:
+
+- `status` (String) Severity of the Security Signal. Valid values are `info`, `low`, `medium`, `high`, `critical`.
+
+Optional:
+
+- `name` (String) Name of the case.
+- `notifications` (List of String) Notification targets for each rule case.
+- `query` (String) A query to associate a third party event to this case.
```
Comment (brett0000FF) [inline docs/resources/security_monitoring_rule.md:217]:
```suggestion
- `query` (String) A query to associate a third-party event to this case.
```

File: docs/resources/service_level_objective.md [hunk 7]
```diff
@@ -79,7 +79,7 @@ resource "datadog_service_level_objective" "bar" {
 
 - `name` (String) Name of Datadog service level objective
 - `thresholds` (Block List, Min: 1) A list of thresholds and targets that define the service level objectives from the provided SLIs. (see [below for nested schema](#nestedblock--thresholds))
-- `type` (String) The type of the service level objective. The mapping from these types to the types found in the Datadog Web UI can be found in the Datadog API [documentation page](https://docs.datadoghq.com/api/v1/service-level-objectives/#create-a-slo-object). Valid values are `metric`, `monitor`.
+- `type` (String) The type of the service level objective. The mapping from these types to the types found in the Datadog Web UI can be found in the Datadog API [documentation page](https://docs.datadoghq.com/api/v1/service-level-objectives/#create-a-slo-object). Valid values are `metric`, `monitor`, `time_slice`.
```
Comment (brett0000FF) [inline docs/resources/service_level_objective.md:82]:
```suggestion
- `type` (String) The type of the service level objective. The mapping from these types to the types found in the Datadog Web UI can be found in the [Datadog API documentation](https://docs.datadoghq.com/api/v1/service-level-objectives/#create-a-slo-object). Valid values are `metric`, `monitor`, `time_slice`.
```

File: docs/resources/security_monitoring_rule.md [hunk 8]
```diff
@@ -134,6 +133,32 @@ Optional:
...
+
+- `default_notifications` (List of String) Notification targets for the default rule case, when none of the third party cases match.
+- `signal_title_template` (String) A template for the signal title; if omitted, the title is generated based on the case name.
+
+<a id="nestedblock--options--third_party_rule_options--root_query"></a>
+### Nested Schema for `options.third_party_rule_options.root_query`
+
+Required:
+
+- `query` (String) Query to filter logs
```
Comment (brett0000FF) [inline docs/resources/security_monitoring_rule.md:154]:
```suggestion
- `query` (String) Query to filter logs.
```

**Review by brett0000FF — APPROVED** [review]
Looks good! Just some nits from Docs.

File: docs/resources/service_level_objective.md (hunk 7 above)
Comment (pietrodll) [reply docs/resources/service_level_objective.md:82]: Thanks for the review! This part of the code isn't really owned by my team, this change was generated automatically following an update of the API client. I believe this PR introduced it: https://github.com/DataDog/datadog-api-client-go/pull/2328. Do you mind if I leave it like this?

File: docs/resources/security_monitoring_rule.md (hunk 1 above)
Comment (pietrodll) [reply docs/resources/security_monitoring_rule.md]: Thanks! Applied in https://github.com/DataDog/terraform-provider-datadog/pull/2231/commits/1e4872ae66b9b61487da7cfbdc524101cda4f968.

File: docs/resources/security_monitoring_rule.md (hunk 2 above)
Comment (pietrodll) [reply docs/resources/security_monitoring_rule.md]: Thanks! Applied in https://github.com/DataDog/terraform-provider-datadog/pull/2231/commits/1e4872ae66b9b61487da7cfbdc524101cda4f968.

File: docs/resources/security_monitoring_rule.md (hunk 3 above)
Comment (pietrodll) [reply docs/resources/security_monitoring_rule.md]: Thanks! Applied in https://github.com/DataDog/terraform-provider-datadog/pull/2231/commits/1e4872ae66b9b61487da7cfbdc524101cda4f968.

File: docs/resources/security_monitoring_rule.md (hunk 4 above)
Comment (pietrodll) [reply docs/resources/security_monitoring_rule.md]: Thanks! Applied in https://github.com/DataDog/terraform-provider-datadog/pull/2231/commits/1e4872ae66b9b61487da7cfbdc524101cda4f968.

File: docs/resources/security_monitoring_rule.md (hunk 5 above)
Comment (pietrodll) [reply docs/resources/security_monitoring_rule.md]: Thanks! Applied in https://github.com/DataDog/terraform-provider-datadog/pull/2231/commits/1e4872ae66b9b61487da7cfbdc524101cda4f968.

File: docs/resources/security_monitoring_rule.md (hunk 6 above)
Comment (pietrodll) [reply docs/resources/security_monitoring_rule.md]: Thanks! Applied in https://github.com/DataDog/terraform-provider-datadog/pull/2231/commits/1e4872ae66b9b61487da7cfbdc524101cda4f968.

File: docs/resources/security_monitoring_rule.md (hunk 8 above)
Comment (pietrodll) [reply docs/resources/security_monitoring_rule.md]: Thanks! Applied in https://github.com/DataDog/terraform-provider-datadog/pull/2231/commits/1e4872ae66b9b61487da7cfbdc524101cda4f968.

**Review by daniel-trinh — DISMISSED** [review]

File: docs/resources/service_level_objective.md (hunk 7 above)
Comment (brett0000FF) [reply docs/resources/service_level_objective.md:82]: You bet. Gotcha, that makes sense. No worries--I think it's fine to leave then. 😄

**Review by brett0000FF — APPROVED** [review]

File: datadog/resource_datadog_security_monitoring_rule.go [hunk 9]
```diff
@@ -665,39 +871,19 @@ func buildCreateStandardPayloadQueries(d *schema.ResourceData) []datadogV2.Secur
...
 
 		if v, ok := query["metrics"]; ok && v != nil {
-			if tfMetrics, ok := v.([]interface{}); ok && len(tfMetrics) > 0 {
-				metrics := make([]string, len(tfMetrics))
-				for i, value := range tfMetrics {
-					metrics[i] = value.(string)
-				}
-				payloadQuery.SetMetrics(metrics)
-			}
+			payloadQuery.SetMetrics(parseStringArray(v.([]interface{})))
```
Comment (nkzou) [inline datadog/resource_datadog_security_monitoring_rule.go:886]: This changes the behavior slightly, in the payload this used to omit metrics if the list was length 0, but now the metrics field gets explicitly set with `"metrics": [],`. Is this intentional?

Comment (pietrodll) [reply datadog/resource_datadog_security_monitoring_rule.go:886]: Good catch, it wasn't really intentional. However I checked the API, it treats empty arrays and omitted fields similarly, so let's keep it like this.

**Review by nkzou — APPROVED** [review]

**Pull request merged**